--- a/net/minecraft/server/ServerScoreboard.java
+++ b/net/minecraft/server/ServerScoreboard.java
@@ -2,7 +2,9 @@
 
 import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
+import java.util.Iterator;
 import java.util.List;
+import java.util.Objects;
 import java.util.Set;
 import javax.annotation.Nullable;
 import net.minecraft.nbt.CompoundTag;
@@ -22,6 +24,7 @@
 import net.minecraft.world.scores.ScoreboardSaveData;
 
 public class ServerScoreboard extends Scoreboard {
+
     private final MinecraftServer server;
     private final Set<Objective> trackedObjectives = Sets.newHashSet();
     private final List<Runnable> dirtyListeners = Lists.newArrayList();
@@ -34,7 +37,7 @@
     public void onScoreChanged(Score score) {
         super.onScoreChanged(score);
         if (this.trackedObjectives.contains(score.getObjective())) {
-            this.server.getPlayerList().broadcastAll(new ClientboundSetScorePacket(ServerScoreboard.Method.CHANGE, score.getObjective().getName(), score.getOwner(), score.getScore()));
+            this.broadcastAll(new ClientboundSetScorePacket(ServerScoreboard.Method.CHANGE, score.getObjective().getName(), score.getOwner(), score.getScore()));
         }
 
         this.setDirty();
@@ -43,7 +46,7 @@
     @Override
     public void onPlayerRemoved(String playerName) {
         super.onPlayerRemoved(playerName);
-        this.server.getPlayerList().broadcastAll(new ClientboundSetScorePacket(ServerScoreboard.Method.REMOVE, (String)null, playerName, 0));
+        this.broadcastAll(new ClientboundSetScorePacket(ServerScoreboard.Method.REMOVE, (String) null, playerName, 0));
         this.setDirty();
     }
 
@@ -51,7 +54,7 @@
     public void onPlayerScoreRemoved(String playerName, Objective objective) {
         super.onPlayerScoreRemoved(playerName, objective);
         if (this.trackedObjectives.contains(objective)) {
-            this.server.getPlayerList().broadcastAll(new ClientboundSetScorePacket(ServerScoreboard.Method.REMOVE, objective.getName(), playerName, 0));
+            this.broadcastAll(new ClientboundSetScorePacket(ServerScoreboard.Method.REMOVE, objective.getName(), playerName, 0));
         }
 
         this.setDirty();
@@ -59,19 +62,20 @@
 
     @Override
     public void setDisplayObjective(DisplaySlot slot, @Nullable Objective objective) {
-        Objective objective2 = this.getDisplayObjective(slot);
+        Objective scoreboardobjective1 = this.getDisplayObjective(slot);
+
         super.setDisplayObjective(slot, objective);
-        if (objective2 != objective && objective2 != null) {
-            if (this.getObjectiveDisplaySlotCount(objective2) > 0) {
-                this.server.getPlayerList().broadcastAll(new ClientboundSetDisplayObjectivePacket(slot, objective));
+        if (scoreboardobjective1 != objective && scoreboardobjective1 != null) {
+            if (this.getObjectiveDisplaySlotCount(scoreboardobjective1) > 0) {
+                this.broadcastAll(new ClientboundSetDisplayObjectivePacket(slot, objective));
             } else {
-                this.stopTrackingObjective(objective2);
+                this.stopTrackingObjective(scoreboardobjective1);
             }
         }
 
         if (objective != null) {
             if (this.trackedObjectives.contains(objective)) {
-                this.server.getPlayerList().broadcastAll(new ClientboundSetDisplayObjectivePacket(slot, objective));
+                this.broadcastAll(new ClientboundSetDisplayObjectivePacket(slot, objective));
             } else {
                 this.startTrackingObjective(objective);
             }
@@ -83,7 +87,7 @@
     @Override
     public boolean addPlayerToTeam(String playerName, PlayerTeam team) {
         if (super.addPlayerToTeam(playerName, team)) {
-            this.server.getPlayerList().broadcastAll(ClientboundSetPlayerTeamPacket.createPlayerPacket(team, playerName, ClientboundSetPlayerTeamPacket.Action.ADD));
+            this.broadcastAll(ClientboundSetPlayerTeamPacket.createPlayerPacket(team, playerName, ClientboundSetPlayerTeamPacket.Action.ADD));
             this.setDirty();
             return true;
         } else {
@@ -91,12 +95,42 @@
         }
     }
 
+    // Paper start
+    public boolean addPlayersToTeam(java.util.Collection<String> players, PlayerTeam team) {
+        boolean anyAdded = false;
+        for (String playerName : players) {
+            if (super.addPlayerToTeam(playerName, team)) {
+                anyAdded = true;
+            }
+        }
+
+        if (anyAdded) {
+            this.broadcastAll(ClientboundSetPlayerTeamPacket.createMultiplePlayerPacket(team, players, ClientboundSetPlayerTeamPacket.Action.ADD));
+            this.setDirty();
+            return true;
+        } else {
+            return false;
+        }
+    }
+    // Paper end
+
     @Override
     public void removePlayerFromTeam(String playerName, PlayerTeam team) {
         super.removePlayerFromTeam(playerName, team);
-        this.server.getPlayerList().broadcastAll(ClientboundSetPlayerTeamPacket.createPlayerPacket(team, playerName, ClientboundSetPlayerTeamPacket.Action.REMOVE));
+        this.broadcastAll(ClientboundSetPlayerTeamPacket.createPlayerPacket(team, playerName, ClientboundSetPlayerTeamPacket.Action.REMOVE));
+        this.setDirty();
+    }
+
+    // Paper start
+    public void removePlayersFromTeam(java.util.Collection<String> players, PlayerTeam team) {
+        for (String playerName : players) {
+            super.removePlayerFromTeam(playerName, team);
+        }
+
+        this.broadcastAll(ClientboundSetPlayerTeamPacket.createMultiplePlayerPacket(team, players, ClientboundSetPlayerTeamPacket.Action.REMOVE));
         this.setDirty();
     }
+    // Paper end
 
     @Override
     public void onObjectiveAdded(Objective objective) {
@@ -108,7 +142,7 @@
     public void onObjectiveChanged(Objective objective) {
         super.onObjectiveChanged(objective);
         if (this.trackedObjectives.contains(objective)) {
-            this.server.getPlayerList().broadcastAll(new ClientboundSetObjectivePacket(objective, 2));
+            this.broadcastAll(new ClientboundSetObjectivePacket(objective, 2));
         }
 
         this.setDirty();
@@ -127,21 +161,21 @@
     @Override
     public void onTeamAdded(PlayerTeam team) {
         super.onTeamAdded(team);
-        this.server.getPlayerList().broadcastAll(ClientboundSetPlayerTeamPacket.createAddOrModifyPacket(team, true));
+        this.broadcastAll(ClientboundSetPlayerTeamPacket.createAddOrModifyPacket(team, true));
         this.setDirty();
     }
 
     @Override
     public void onTeamChanged(PlayerTeam team) {
         super.onTeamChanged(team);
-        this.server.getPlayerList().broadcastAll(ClientboundSetPlayerTeamPacket.createAddOrModifyPacket(team, false));
+        this.broadcastAll(ClientboundSetPlayerTeamPacket.createAddOrModifyPacket(team, false));
         this.setDirty();
     }
 
     @Override
     public void onTeamRemoved(PlayerTeam team) {
         super.onTeamRemoved(team);
-        this.server.getPlayerList().broadcastAll(ClientboundSetPlayerTeamPacket.createRemovePacket(team));
+        this.broadcastAll(ClientboundSetPlayerTeamPacket.createRemovePacket(team));
         this.setDirty();
     }
 
@@ -150,7 +184,11 @@
     }
 
     protected void setDirty() {
-        for(Runnable runnable : this.dirtyListeners) {
+        Iterator iterator = this.dirtyListeners.iterator();
+
+        while (iterator.hasNext()) {
+            Runnable runnable = (Runnable) iterator.next();
+
             runnable.run();
         }
 
@@ -158,16 +196,25 @@
 
     public List<Packet<?>> getStartTrackingPackets(Objective objective) {
         List<Packet<?>> list = Lists.newArrayList();
+
         list.add(new ClientboundSetObjectivePacket(objective, 0));
+        DisplaySlot[] adisplayslot = DisplaySlot.values();
+        int i = adisplayslot.length;
 
-        for(DisplaySlot displaySlot : DisplaySlot.values()) {
-            if (this.getDisplayObjective(displaySlot) == objective) {
-                list.add(new ClientboundSetDisplayObjectivePacket(displaySlot, objective));
+        for (int j = 0; j < i; ++j) {
+            DisplaySlot displayslot = adisplayslot[j];
+
+            if (this.getDisplayObjective(displayslot) == objective) {
+                list.add(new ClientboundSetDisplayObjectivePacket(displayslot, objective));
             }
         }
 
-        for(Score score : this.getPlayerScores(objective)) {
-            list.add(new ClientboundSetScorePacket(ServerScoreboard.Method.CHANGE, score.getObjective().getName(), score.getOwner(), score.getScore()));
+        Iterator iterator = this.getPlayerScores(objective).iterator();
+
+        while (iterator.hasNext()) {
+            Score scoreboardscore = (Score) iterator.next();
+
+            list.add(new ClientboundSetScorePacket(ServerScoreboard.Method.CHANGE, scoreboardscore.getObjective().getName(), scoreboardscore.getOwner(), scoreboardscore.getScore()));
         }
 
         return list;
@@ -175,10 +222,17 @@
 
     public void startTrackingObjective(Objective objective) {
         List<Packet<?>> list = this.getStartTrackingPackets(objective);
+        Iterator iterator = this.server.getPlayerList().getPlayers().iterator();
 
-        for(ServerPlayer serverPlayer : this.server.getPlayerList().getPlayers()) {
-            for(Packet<?> packet : list) {
-                serverPlayer.connection.send(packet);
+        while (iterator.hasNext()) {
+            ServerPlayer entityplayer = (ServerPlayer) iterator.next();
+            if (entityplayer.getBukkitEntity().getScoreboard().getHandle() != this) continue; // CraftBukkit - Only players on this board
+            Iterator iterator1 = list.iterator();
+
+            while (iterator1.hasNext()) {
+                Packet<?> packet = (Packet) iterator1.next();
+
+                entityplayer.connection.send(packet);
             }
         }
 
@@ -187,11 +241,16 @@
 
     public List<Packet<?>> getStopTrackingPackets(Objective objective) {
         List<Packet<?>> list = Lists.newArrayList();
+
         list.add(new ClientboundSetObjectivePacket(objective, 1));
+        DisplaySlot[] adisplayslot = DisplaySlot.values();
+        int i = adisplayslot.length;
 
-        for(DisplaySlot displaySlot : DisplaySlot.values()) {
-            if (this.getDisplayObjective(displaySlot) == objective) {
-                list.add(new ClientboundSetDisplayObjectivePacket(displaySlot, objective));
+        for (int j = 0; j < i; ++j) {
+            DisplaySlot displayslot = adisplayslot[j];
+
+            if (this.getDisplayObjective(displayslot) == objective) {
+                list.add(new ClientboundSetDisplayObjectivePacket(displayslot, objective));
             }
         }
 
@@ -200,10 +259,17 @@
 
     public void stopTrackingObjective(Objective objective) {
         List<Packet<?>> list = this.getStopTrackingPackets(objective);
+        Iterator iterator = this.server.getPlayerList().getPlayers().iterator();
 
-        for(ServerPlayer serverPlayer : this.server.getPlayerList().getPlayers()) {
-            for(Packet<?> packet : list) {
-                serverPlayer.connection.send(packet);
+        while (iterator.hasNext()) {
+            ServerPlayer entityplayer = (ServerPlayer) iterator.next();
+            if (entityplayer.getBukkitEntity().getScoreboard().getHandle() != this) continue; // CraftBukkit - Only players on this board
+            Iterator iterator1 = list.iterator();
+
+            while (iterator1.hasNext()) {
+                Packet<?> packet = (Packet) iterator1.next();
+
+                entityplayer.connection.send(packet);
             }
         }
 
@@ -212,9 +278,13 @@
 
     public int getObjectiveDisplaySlotCount(Objective objective) {
         int i = 0;
+        DisplaySlot[] adisplayslot = DisplaySlot.values();
+        int j = adisplayslot.length;
 
-        for(DisplaySlot displaySlot : DisplaySlot.values()) {
-            if (this.getDisplayObjective(displaySlot) == objective) {
+        for (int k = 0; k < j; ++k) {
+            DisplaySlot displayslot = adisplayslot[k];
+
+            if (this.getDisplayObjective(displayslot) == objective) {
                 ++i;
             }
         }
@@ -227,17 +297,31 @@
     }
 
     private ScoreboardSaveData createData() {
-        ScoreboardSaveData scoreboardSaveData = new ScoreboardSaveData(this);
-        this.addDirtyListener(scoreboardSaveData::setDirty);
-        return scoreboardSaveData;
+        ScoreboardSaveData persistentscoreboard = new ScoreboardSaveData(this);
+
+        Objects.requireNonNull(persistentscoreboard);
+        this.addDirtyListener(persistentscoreboard::setDirty);
+        return persistentscoreboard;
     }
 
     private ScoreboardSaveData createData(CompoundTag nbt) {
         return this.createData().load(nbt);
     }
 
+    // CraftBukkit start - Send to players
+    private void broadcastAll(Packet packet) {
+        for (ServerPlayer entityplayer : (List<ServerPlayer>) this.server.getPlayerList().players) {
+            if (entityplayer.getBukkitEntity().getScoreboard().getHandle() == this) {
+                entityplayer.connection.send(packet);
+            }
+        }
+    }
+    // CraftBukkit end
+
     public static enum Method {
-        CHANGE,
-        REMOVE;
+
+        CHANGE, REMOVE;
+
+        private Method() {}
     }
 }
